core/src/main/java/org/infinispan/CacheStream.java

```diff
@@ -0,0 +1,251 @@
+ * distribution controlled by the segments that each key maps to.  All intermediate operations are lazy, even the
+ * special cases described in later paragraphs and are not evaluated until a final terminal operation is invoked on
+ * the stream.  Essentially each set of intermediate operations is shipped to each remote node where they are applied
+ * to a local stream there and finally the terminal operation is completed.  If this stream is parallel the processing
+ * on remote nodes is also done using a parallel stream.</p>
+ *
+ * <p>Parallel distribution is enabled by default for all operations except for {@link CacheStream#iterator()} &
+ * {@link CacheStream#spliterator()}.  Please see {@link CacheStream#sequentialDistribution()} and
+ * {@link CacheStream#parallelDistribution()}.  With this disabled only a single node will process the operation
+ * at a time (includes locally).</p>
+ *
+ * <p>Rehash aware is enabled by default for all operations which will provide guaranteed consistency for all operations
+ * except for {@link CacheStream#forEach(Consumer)}.  Please the method above for details about its consistency
+ * guarantees.  If you wish to disable rehash aware operations you can disable them by calling
+ * {@link CacheStream#disableRehashAware()} which should provide better performance for some operations.  The
+ * performance is most affected for the key aware operations {@link CacheStream#iterator()},
+ * {@link CacheStream#spliterator()}, {@link CacheStream#forEach(Consumer)}</p>
+ *
+ * <p>Some terminal operators are special in that they act like an intermediate iterator operation.  That is that
+ * it is an intermediate operation, but it requires processing the results using an interator intermediately before
+ * the stream can complete.</p>
+ *
+ * <p>A good example of an intermediate iterator operation is using distinct intermediate operation. What will happen
+ * is upon calling the terminal operation an iterator operation will be ran using all of
+ * the intermediate operations up to the distinct operation remotely.  This iterator is then used to fuel a local
+ * stream where all of the remaining intermediate operations are performed and then finally the terminal operation is
+ * applied as normal.  Note in this case the intermediate iterator still obeys the
+ * {@link CacheStream#distributedBatchSize(int)} setting irrespective of the terminal operator.</p>
+ *
+ * @param <R> The type of the stream
+ * @since 8.0
+ */
+public interface CacheStream<R> extends Stream<R> {
+   /**
+    * This would disable sending requests to all other remote nodes compared to one at a time. This can reduce memory
+    * pressure on the originator node at the cost of performance.
+    * <p>Parallel distribution is enabled by default except for {@link CacheStream#iterator()} &
+    * {@link CacheStream#spliterator()}</p>
+    * @return This stream again with parallel distribution disabled
```

Comment:
`This stream again` sounds too specific, like you're guaranteeing always to return the same instance.